src/actions/app.cr

```diff
@@ -49,86 +50,174 @@ module DppmRestApi::Actions::App
-  # start the service associated with the given application
-  relative_put "/:app_name/service/start" do |context|
+
+  {% for action in ["reload", "restart", "start", "stop"] %}
+  # {[action.id]} the service associated with the given application
+  relative_put "/:app_name/service/{{action.id}}" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Update
-    # TODO: start the service
+    app = Actions.prefix.new_app(context.params.url["app_name"])
+    app.service? || app.service_create(database_name: context.params.query["database_name"]?)
+    app.service.{{action.id}}
   end
+  
+  {% end %}
   # get the status of the service associated with the given application
   relative_get "/:app_name/service/status" do |context|
+    app_name = context.params.url["app_name"]
     raise Unauthorized.new context unless Actions.has_access? context, Access::Read
     # TODO: get the status of the service
   end
-  # stop the service associated with the given application
-  relative_put "/:app_name/service/stop" do |context|
-    raise Unauthorized.new context unless Actions.has_access? context, Access::Update
-    # TODO: stop the service
-  end
+
   # lists dependent library packages
   relative_get "/:app_name/libs" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Read
     # TODO: list dependencies
   end
+
   # return the base application package
   relative_get "/:app_name/app" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Read
-    # TODO: return the base application package
+    app_name = context.params.url["app_name"]
+    build_json context.response do |builder|
+      builder.field(app_name) do
+        Actions.prefix.new_app(app_name).to_json builder
+      end
+    end
   end
+
   # returns information present in pkg.con as JSON
   relative_get "/:app_name/pkg" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Read
-    # TODO: return package data
+    app_name = context.params.url["app_name"]
+    build_json context.response do |builder|
+      builder.field(app_name) do
+        Actions.prefix.new_app(context.params.url["app_name"]).pkg.to_json builder
+      end
+    end
   end
+
   # if the `"stream"` query parameter is set, attempt to upgrade to a websocket
   # and stream the results. Otherwise return a JSON-formatted output of the
   # current log data.
   relative_get "/:app_name/logs" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Read
-    # TODO: upgrade to websocket or output logs to date
+    app_name = context.params.url["app_name"]
+    app = Actions.prefix.new_app app_name
+    stream_type = context.params.query["stream_type"]? || [] of String
+    stream_type = if stream_type.empty?
+                    ["errors", "output"]
+                  else
+                    ["errors", "output"] & stream_type
```

Comment:
They aren't the only streams possible, and we shouldn't know about the .log extension (I push a fix upstream)

Reply:
Are all possible streams knowable or should we simply not filter that?

Reply:
This was `App#each_log_file`, now `App#each_log_stream` in master